pkg/action/executor/rescheduler.go

```diff
@@ -0,0 +1,261 @@
+package executor
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+
+	"github.com/vmturbo/kubeturbo/pkg/action/turboaction"
+	"github.com/vmturbo/kubeturbo/pkg/action/util"
+	"github.com/vmturbo/kubeturbo/pkg/discovery/probe"
+	"github.com/vmturbo/kubeturbo/pkg/turbostore"
+
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"github.com/golang/glog"
+)
+
+type ReScheduler struct {
+	kubeClient *client.Client
+	broker     turbostore.Broker
+}
+
+func NewReScheduler(client *client.Client, broker turbostore.Broker) *ReScheduler {
+	return &ReScheduler{
+		kubeClient: client,
+		broker:     broker,
+	}
+}
+
+func (r *ReScheduler) Execute(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction, error) {
+	if actionItem == nil {
+		return nil, errors.New("ActionItem passed in is nil")
+	}
+	action, err := r.buildPendingReScheduleTurboAction(actionItem)
+	if err != nil {
+		return nil, err
+	}
+	return r.reSchedule(action)
+}
+
+func (r *ReScheduler) buildPendingReScheduleTurboAction(actionItem *proto.ActionItemDTO) (*turboaction.TurboAction,
+	error) {
+	// Find out the pod to be re-scheduled.
+	targetPod := actionItem.GetTargetSE()
+	podIdentifier := targetPod.GetId() // podIdentifier must have format as "Namespace:Name"
+	originalPod, err := util.GetPodFromCluster(r.kubeClient, podIdentifier)
+	if err != nil {
+		return nil, fmt.Errorf("Try to move pod %s, but could not find it in the cluster.", podIdentifier)
+	}
+
+	// Find out where to re-schedule the pod.
+	newSEType := actionItem.GetNewSE().GetEntityType()
+	if newSEType != proto.EntityDTO_VIRTUAL_MACHINE && newSEType != proto.EntityDTO_PHYSICAL_MACHINE {
+		return nil, errors.New("The target service entity for move destiantion is neither a VM nor a PM.")
+	}
+	targetNode := actionItem.GetNewSE()
+
+	var machineIPs []string
+	switch newSEType {
+	case proto.EntityDTO_VIRTUAL_MACHINE:
+		// K8s uses Ip address as the Identifier. The VM name passed by actionItem is the display name
+		// that discovered by hypervisor. So here we must get the ip address from virtualMachineData in
+		// targetNode entityDTO.
+		vmData := targetNode.GetVirtualMachineData()
+		if vmData == nil {
+			return nil, errors.New("Missing VirtualMachineData in ActionItemDTO from server")
+		}
+		machineIPs = vmData.GetIpAddress()
+		break
+	case proto.EntityDTO_PHYSICAL_MACHINE:
+		// TODO
+		// machineIPS = <valid physical machine IP>
```

Comment:
What is the gap here to support PM? We need to have a list that lists the gap between VM support and bare metal support, so that when we have a customer using bare metal, we know what is supported and what is not.